kaggle_environments/envs/lux_ai_2022/luxai2022/env.py

```diff
@@ -0,0 +1,618 @@
+        else:
+            if self.env_steps == 0:
+                return get_act_space_init(config=self.env_cfg, agent=agent)
+            return get_act_space(self.state.units, self.state.factories, config=self.env_cfg, agent=agent)
+
+    def render(self, mode="human"):
+        """
+        Renders the environment. In human mode, it can print to terminal, open
+        up a graphical window, or open up some other display that a human can see and understand.
+        """
+        # if len(self.agents) == 2:
+        #     string = ("Current state: Agent1: {} , Agent2: {}".format(MOVES[self.state[self.agents[0]]], MOVES[self.state[self.agents[1]]]))
+        # else:
+        #     string = "Game over"
+
+        if mode == "human":
+            if self.py_visualizer is None:
+                self.py_visualizer = Visualizer(self.state)
+            self.py_visualizer.update_scene(self.state)
+            self.py_visualizer.render()
+
+    def close(self):
+        """
+        Close should release any graphical displays, subprocesses, network connections
+        or any other environment data which should not be kept around after the
+        user is no longer using the environment.
+        """
+        pass
+
+    def get_state(self):
+        return self.state
+
+    def set_state(self, state: State):
+        self.state = state
+        self.env_steps = state.env_steps
+        self.seed_rng = state.seed_rng
+        self.seed = state.seed
+        # TODO - throw warning if setting state from a different configuration than initialized with
+        self.env_cfg = state.env_cfg
+        self.max_episode_length = self.env_cfg.max_episode_length
+
+    def reset(self, seed=None):
+        """
+        Reset needs to initialize the `agents` attribute and must set up the
+        environment so that render(), and step() can be called without issues.
+
+        Here it initializes the `num_moves` variable which counts the number of
+        hands that are played.
+
+        Returns the observations for each agent
+        """
+        seed_rng = np.random.RandomState(seed=seed)
+        self.agents = self.possible_agents[:]
+        self.env_steps = 0
+        self.seed = seed
+        board = Board(seed=seed, env_cfg=self.env_cfg)
+        weather_schedule = weather.generate_weather_schedule(seed_rng, self.state.env_cfg)
+        self.state: State = State(seed_rng=seed_rng, seed=seed, env_cfg=self.state.env_cfg, env_steps=0, board=board, weather_schedule=weather_schedule)
+        self.max_episode_length = self.env_cfg.max_episode_length
+        for agent in self.possible_agents:
+            self.state.units[agent] = OrderedDict()
+            self.state.factories[agent] = OrderedDict()
+        obs = self.state.get_obs()
+        observations = {agent: obs for agent in self.agents}
+        return observations
+
+    def step(self, actions):
```

Comment:
given that python doesn't scope variables it might be better to move the actual logic of each of these steps into their own function

```
def step(self, actions):
    if early_game:
        if bidding:
            bid_result = bidding(actions)
        else:
             placement_result = factory_placement(actions)
      else:
          validation_result = validate_actions(actions)
         etc...
```
this would protect against an errant `unit_a` being defined on line 190 being used on line 400 by accident (I've done this too many times in python)

Reply:
refactoring!